Refactor and improve gconftool2 module

azaghal opened the pull request:
##### SUMMARY

While trying out the gconftool2 module I ran into a number of issues:

- Commands were run via `subprocess.Popen` in combination with `shell=True`. This caused issues when I tried to set a keyboard shortcut for Guake to `<Shift>Left` (`<>` acting as input/output redirections). All other special shell characters would cause issues as well.
- Module did not handle correctly situation of detecting errors in case where a key does not exist, resulting in failures.
- The `get` state required that parameters `value` and `value_type` are provided (even though it makes no sense for this state, since you want to read the value, not set it).
- Module was setting Ansible facts instead of return value using values retrieved from GConf. This was done for all operations (while, technically, the `present` state is the only one making sense).

The code was rewritten to fix all the issues listed above, and cleaned-up along the way.

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

`gconftool2`

##### ANSIBLE VERSION

```
ansible 2.4.0 (gconftool2-refactor 692ce680b6) last updated 2017/03/24 20:45:39 (GMT +200)
  config file = 
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Dec 13 2016, 14:29:11) [GCC 4.9.3]
```

##### ADDITIONAL INFORMATION

There are some caveats with the use of `gconftool-2` CLI tool. Namely:

- Some settings can be set either system-wide (defaults) or by user. If value is the same, it is not possible to detect if the user has set the value or not locally, making it impossible to detect changes in cases where the key is removed (state `absent`).
- It does not return proper error codes. Therefore one must rely on processing standard error.
- Standard error *may* be prone to user localisation, in which case the error check for `get` operation will not be reliable.
- There is a Python GConf module that could be used instead of running commands on the system. Not sure if it would be worth it to introduce dependency, but it may solve some of the limitations of `gconftool-2` listed above.

azaghal commented:
Anything I could do to get this pull request rolling?

Akasurde commented:
@azaghal Could you please do rebase on given branch? I think the file changed since you last posted.

azaghal commented:
Did a rebase. Mind you, this pull request is orthogonal to the last commit to the module (since it's a rewrite of most of the things).

Akasurde commented on lib/ansible/modules/system/gconftool2.py:
```diff
@@ -197,74 +285,28 @@ def main():
-
-    module.exit_json(changed=change, ansible_facts=facts)
+    # If direct access was requested, config_source must be provided.
+    if module.params['direct'] and (module.params['config_source'] is None):
```
Could you please use `required_if` something [like](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/network/eos/eos_banner.py#L156) ?

azaghal replied:
I'm guessing the correct thing to use here would be `require_together`?

alikins replied:
If config_source is only used in combo with direct, whichs looks to be the case, it may make sense to collapse it into one arg (config_source). Then GConfTool2.__init__ could:

``` python

if config_source:
   self.base_command += ["--direct", "--config-source", config_source]

```

That assumes there won't be a scenario where it makes sense to call 'gconftool-2 --direct' or 'gconftool-2 --config-souce some_config_source' independently.

azaghal replied:
The main reason to keep it was backwards compatibility parameter-wise. I could break this as  well I guess (just to figure out if module should keep name or be spined of as another module.

azaghal replied:
I'll go ahead and remove the `direct` parameter - `config_source` will involve using `--direct`.

Akasurde commented on lib/ansible/modules/system/gconftool2.py:
```diff
@@ -197,74 +285,28 @@ def main():
+        module.fail_json(msg='If "direct" is "yes", then the "config_source" must be specified')
+
+    # If present state was specified, value and value type must be provided.
+    if module.params['state'] == 'present' and (module.params['value'] is None or module.params['value_type'] is None):
```
ditto as `required_if`

azaghal replied:
I'm guessing the correct thing to use here would be `require_together`?

alikins replied:
I think for this case, this should work:

``` python

required_if = [('state', 'present', ('value', 'value_type'))]
```

But haven't tested that...

azaghal replied:
Seems to work (tm). I'll update relevant sections of code.

Akasurde commented on lib/ansible/modules/system/gconftool2.py:
```diff
@@ -36,146 +38,233 @@
+    manipulating a single preference in GConf repository.
+    """
+
+    def __init__(self, module, direct=False, config_source=None, check_mode=False):
```
why not take only `module` as parameter, we can derive all values from it.

azaghal replied:
Wanted to preserve a more loose coupling between the class handling actions and Ansible module object itself. I.e. handle parameter checks and processing etc within the `main()` function, and only pass-in the necessary parameters to the class doing the actual work.

What is the usual convention amongst existing modules?

azaghal replied:
Seems to be a mixed bag. `hg`, `bzr` use same approach of getting params out in `main()`, `service` takes in just the module object. Looks like 215 with just module in `__init__` vs 100 with other params as well, based on some quick grepping:

```
grep -r __init__ lib/ansible/modules/  | grep module | grep  'module)' | wc -l
grep -r __init__ lib/ansible/modules/  | grep module | grep -v 'module)' | wc -l
```

azaghal replied:
Asked about this on IRC too, seems there's no strong preferences, so I'll keep it as-is for now.

Akasurde commented on lib/ansible/modules/system/gconftool2.py:
```diff
@@ -36,146 +38,233 @@
+        """
+
+        self.module = module
+        self.base_command = ["gconftool-2"]
```
Use [get_bin_path](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L1908) to get path of file

azaghal replied:
Will fix, didn't know about that one, looks very useful :)

Akasurde commented on lib/ansible/modules/system/gconftool2.py:
```diff
@@ -36,146 +38,233 @@
+                       If enabled, valid config_source must be passed in as well.
+        :type direct: bool
+
+        :param config_source: Configuratrion source to use in case direct access has been requested.
```
s/Configuratrion/Configuration/

azaghal replied:
Will fix.

Akasurde commented on lib/ansible/modules/system/gconftool2.py:
```diff
@@ -36,146 +38,233 @@
+            return True
+
+        # Set-up command to run.
+        command = list(self.base_command)
```
`self.base_command` is already list so I think this is useless.

azaghal replied:
The intention is to create copy of `base_command` - which I keep intact since both `set` and `unset` use it as basis for constructing relevant commands, and both `set` and `unset` will call the `read` method at least once. .

Akasurde commented on lib/ansible/modules/system/gconftool2.py:
```diff
@@ -197,74 +285,28 @@ def main():
+    if module.params['state'] == 'get':
+        value = gconftool2.get(module.params['key'])
+        module.exit_json(changed=False, value=value)
+    elif module.params['state'] == 'present':
```
no need to check for present as it is default state

azaghal replied:
I think it improves readability of the code in this particular case. If you really think it should be removed, though, I will do that.

alikins replied:
@azaghal I would leave it as you have it, with the explicit check for 'present'

Akasurde commented on lib/ansible/modules/system/gconftool2.py:
```diff
@@ -197,74 +285,28 @@ def main():
+        module.exit_json(changed=changed)
+    elif module.params['state'] == 'absent':
+        changed = gconftool2.unset(module.params['key'])
+        module.exit_json(changed=changed)
```
You are not returning ansible_facts here. This will break backward compatibility.

azaghal replied:
This was done on purpose, but I would like to get some feedback (and will test it again) to make sure I understood things correctly (at least initially when I had look at it it made sense, but it was some time ago).

The old implementation does a `module.exit_json(changed=changed, ansible_facts=facts)`, where the `facts` is a dictionary with one key - `gconftool2`. It was my understanding at the time that by doing this you set globally available facts that will be available for subsequent tasks in the same play (unconditionally I might add). I considered that counter-intuitive since I do not think this module fits the category of general-purpose fact reading module. Instead I tried to make the value available for the `get` state via `register` keyword in task definition.

Perhaps to illustrate things a bit better, this is my understanding how the old code worked (keep in mind I might be completely wrong, I just recall testing something like this back then :) ):

```
- name: Read the value of widget font
  gconftool2:
    key: "/desktop/gnome/interface/font_name"
    state: get
# Notice global presence of variable no matter what.
- debug: var=gconftool2
```

In case of new module you would need to be more explicit that you want the value:
```
- name: Read the value of widget font
  gconftool2:
    key: "/desktop/gnome/interface/font_name"
    state: get
  register: fontname
- debug: var=fontname
```

azaghal replied:
Ok, I can confirm my rationale for breaking backwards compatibility above. When using `ansible_facts=facts`, you end-up merging the values into Ansible facts, which I consider wrong in respect with this module - since it's not one of the generic `*_facts` modules. The value changes on each invocation, overwriting what was previously set, it gives limited view of what values are available (just one, the last one changed/read etc), whereas the `*_facts` modules are usually (I think?) invoked once, and they provide a complete picture of facts they cover.

E.g., if you invoke `xattr` module, you need to explicitly `register` the variable in a task, and it is close behaviour-wise to `gconftool2`.

I do not see why `gconftool2` should receive special treatment in that regard, especially since it does not behave as a kind of "facts" module (like `setup` and similar).

Although, I do store less information in my refactoring as well - so that's another thing to discuss after we settle on backwards compatibility decision.

alikins replied:
@azaghal I think I agree with you that the old code was kind of doing the wrong thing by settings facts at all.
And doubly so if:

> the value changes on each invocation, overwriting what was previously set, it gives limited view of what values are available (just one, the last one changed/read etc),

Not sure what to do about the backwards compatibility though. Suppose we could:

- deprecated gconftool2 module
- add this module as new module  (gconftool2_ctl?)
- possibly add another new module for facts (gconftool2_facts)
   - and fix up what it returns for facts so they dont clobber 
        - would each config_source need a namespace in this case? somthing like
``` python
            {
                'gconftool_facts_default': {'some_key': (some_value, some_value_type)},
                'gconftool_facts_some_other_config_source': {'some_key': (some_other_value, some_value_type)}
            }
```

azaghal commented:
Could I get feedback on some of the comments left in the review? Main blocker I can see is the breaking of backwards compatibility for which I've provided explanation. Of course, if it's really paramount that same fact-setting is preserved, I will redo that part of code to match existing implementation. I would just like to point out that as long as module is in preview, it might be a bit easier "social-contract-wise" to make it behave better.

alikins commented on lib/ansible/modules/system/gconftool2.py:
```diff
@@ -36,146 +38,233 @@
+        command += ["--get", key]
+
+        # Run the command and fetch standard return code, stdout, and stderr.
+        rc, out, err = self.module.run_command(command)
```
Could try running the command with the env setup to use LANG=C 

You can specify the env for the command either with the 'environ_update' arg to run_command() or
by setting module.run_command_environ_update before the run command.

azaghal replied:
Ah, that's an excellent point. I'll take care of it in next batch of commits.

sivel commented:
Re-triggering CI

azaghal committed:
Major refactoring of the gconftool2 module:

- Fixed indentation for a number of items in inline module documentation.
- Added more examples to module documentation.
- Rewrote most of the code handling changes and retrival of values from GConf
  repository.
    - Refactored module to use AnsibleModule.run_command instead of
      subprocess.Popen.
    - Properly assemble command to run using list, and avoid running via shell
      intepreter. This should allow shell-special characters in values (<>|&).
    - An empty value can be provided now.
    - Default state is "present" to keep it consistent with the rest of modules.
    - Handle some corner-cases for detecting failures better (do not fail on
      non-existing key).
    - Do not set Ansible facts in the module.
    - Provide return value only for the "get" state.
    - Be more conservative when stripping characters from gconftool-2 output in
      order to avoid loss of data.
    - Removed parameter direct. It is sufficient to provide config_source.
    - Set environment variable LANG to C in order to avoid error detection
      issues related to localisation.

azaghal commented:
Applied some fixes, and rebased the branch. If backwards compatibility is still considered to be the blocker after explanations posted, do let me know.
